### PR TITLE
cache: Add note about buffer alignment and size

### PR DESCRIPTION
**Pull request by carlocaione**
Cache operations that act on unaligned buffers (with respect to cache line) or buffers that are not multiple of the cache line size are at least dangerous with an high risk of data and memory corruption.

While we do not enforce a contract on the buffer characteristics between users and APIs, we can at least add a note in the documentation specifying an undefined behaviour when the passed buffer is unaligned or wrongly sized.

### CODE REVIEW

**Review by gmarull — DISMISSED** [review]

File: include/zephyr/arch/cache.h [hunk 1]
```diff
@@ -89,6 +89,9 @@ extern int arch_dcache_flush_and_invd_all(void);
  *
  * Flush the specified address range of the data cache.
  *
+ * @note: when @p addr is not aligned to the cache line and/or @p size is not
```
Comment (gmarull) [inline include/zephyr/arch/cache.h:92]:
```suggestion
 * @note when @p addr is not aligned to the cache line and/or @p size is not
```

forgot this one (applies to all others)

**Review by nashif — DISMISSED** [review]

File: include/zephyr/arch/cache.h [hunk 2]
```diff
@@ -105,6 +108,9 @@ extern int arch_dcache_flush_range(void *addr, size_t size);
  *
  * Invalidate the specified address range of the data cache.
  *
+ * @note when @p addr is not aligned to the cache line and/or @p size is not
```
Comment (hubertmis) [inline include/zephyr/arch/cache.h:111]: I believe for flushing (or writing back) the behavior for unaligned buffer could be defined: it's writing back the passed structure and all other data structures sharing cache lines with the passed one.

Any module can expect a write-back of its data structure triggered by hardware anytime (during line evictions), so triggering it by other software modules should be fine.

If we don't define this behavior, we would enforce aligning and padding data structures for which it's not strictly necessary. Resulting in increased memory consumption.

File: include/zephyr/arch/cache.h [hunk 3]
```diff
@@ -121,6 +127,9 @@ extern int arch_dcache_invd_range(void *addr, size_t size);
  *
  * Flush and Invalidate the specified address range of the data cache.
  *
+ * @note when @p addr is not aligned to the cache line and/or @p size is not
```
Comment (hubertmis) [inline include/zephyr/arch/cache.h:130]: I would say that invalidating data structures sharing cache lines with other data structures, which can be written by any software module can lead to data corruption and the behavior is undefined.

If an invalidated data structure shares cache lines with read-only data, then it's fine.

File: include/zephyr/arch/cache.h [hunk 4]
```diff
@@ -89,6 +89,9 @@ extern int arch_dcache_flush_and_invd_all(void);
  *
  * Flush the specified address range of the data cache.
  *
+ * @note when @p addr is not aligned to the cache line and/or @p size is not
```
Comment (hubertmis) [inline include/zephyr/arch/cache.h:92]: For flush+invalidate applies the same comment I have for flush-only functions.

**Review by hubertmis — COMMENTED** [review]
I think these notes should be more precise to avoid unwanted side effects of too-strict alignment and padding of data structures by users of these functions.

**Review by gmarull — APPROVED** [review]

**Review by hubertmis — APPROVED** [review]

**Pull request merged**